src/vcpkg/base/cmd-parser.cpp

```diff
@@ -82,4 +324,692 @@ namespace vcpkg
+            }
+        }
+
+        return found > 0;
+    }
+
+    bool CmdParser::parse_switch(StringView switch_name, StabilityTag stability, Optional<bool>& value)
+    {
+        bool target;
+        bool parsed = parse_switch(switch_name, stability, target);
+        if (parsed)
+        {
+            value.emplace(target);
+        }
+
+        return parsed;
+    }
+
+    bool CmdParser::parse_switch(StringView switch_name, StabilityTag stability)
+    {
+        bool target = false;
+        parse_switch(switch_name, stability, target);
+        return target;
+    }
+
+    bool CmdParser::parse_switch(StringView switch_name,
+                                 StabilityTag stability,
+                                 bool& value,
+                                 const LocalizedString& help_text)
+    {
+        Checks::check_exit(VCPKG_LINE_INFO, stability != StabilityTag::ImplementationDetail);
+        options_table.emplace(SwitchName{switch_name.to_string(), stability}, help_text);
+        return parse_switch(switch_name, stability, value);
+    }
+
+    bool CmdParser::parse_switch(StringView switch_name,
+                                 StabilityTag stability,
+                                 Optional<bool>& value,
+                                 const LocalizedString& help_text)
+    {
+        Checks::check_exit(VCPKG_LINE_INFO, stability != StabilityTag::ImplementationDetail);
+        options_table.emplace(SwitchName{switch_name.to_string(), stability}, help_text);
+        return parse_switch(switch_name, stability, value);
+    }
+
+    bool CmdParser::parse_switch(StringView switch_name, StabilityTag stability, const LocalizedString& help_text)
+    {
+        Checks::check_exit(VCPKG_LINE_INFO, stability != StabilityTag::ImplementationDetail);
+        options_table.emplace(SwitchName{switch_name.to_string(), stability}, help_text);
+        return parse_switch(switch_name, stability);
+    }
+
+    bool CmdParser::parse_option(StringView option_name, StabilityTag stability, std::string& value)
+    {
+        std::size_t found = 0;
+        for (std::size_t idx = 0; idx < argument_strings.size(); ++idx)
+        {
+            if (argument_parsed[idx] != false)
+            {
+                continue;
+            }
+
+            auto parse_result =
+                try_parse_option(argument_strings_lowercase[idx], argument_strings[idx], option_name, stability, value);
+            if (parse_result == TryParseOptionResult::NoMatch)
+            {
+                continue;
+            }
+
+            if (found == 1)
+            {
+                errors.emplace_back(msg::format_error(msgOptionUsedMultipleTimes, msg::option = option_name));
+            }
+
+            if (parse_result == TryParseOptionResult::ValueIsNextParameter)
+            {
+                argument_parsed[idx] = true;
+                if (idx + 1 == argument_strings.size() || argument_parsed[idx + 1] != false)
```

Comment:
I'm having a hard time understanding a case where `argument_parsed[idx+1]` could be `true` here. Do you have a test case for this?

Reply:
I am pretty sure this is covered in other tests but I added `"Options do not return consumed args"` just in case.

This comment also convinced me to change the behavior when parsing multi_options to always consume them all and blow away the by-ref parameter, see changes in a7e1a4e3528f24f035150841c37648c485b528fb